src/blueapi/plugins/data_writing.py

```diff
@@ -0,0 +1,106 @@
+
+    @abstractmethod
+    async def update(self) -> None:
+        ...
+
+
+class ServiceDataCollectionProvider(DataCollectionProvider):
+    _collection_group: str
+    _current_collection: Optional[DataCollection]
+
+    def __init__(self, collection_group: str) -> None:
+        self._collection_group = collection_group
+        self._current_collection = None
+
+    @property
+    def current_data_collection(self) -> Optional[DataCollection]:
+        return self._current_collection
+
+    async def update(self) -> None:
+        async with ClientSession() as session:
+            async with session.post(
+                f"http://localhost:8089/collection/{self._collection_group}"
+            ) as response:
+                if response.status == 200:
+                    json = await response.json()
+                    result = DataCollectionSetupResult.parse_obj(json)
+                else:
+                    raise Exception(response.status)
+        if result.directories_created:
+            result.collection
+        else:
+            raise Exception()
+
+
+class InMemoryDataCollectionProvider(DataCollectionProvider):
+    _collection_group: str
+    _scan_number: itertools.count
+    _current_collection: Optional[DataCollection]
+
+    def __init__(self, collection_group: str) -> None:
+        self._collection_group = collection_group
+        self._scan_number = itertools.count()
+        self._current_collection = None
+
+    @property
+    def current_data_collection(self) -> Optional[DataCollection]:
+        return self._current_collection
+
+    async def update(self) -> None:
+        scan_number = next(self._scan_number)
+        self._current_collection = DataCollection(
+            collection_number=scan_number,
+            group=self._collection_group,
+            raw_data_files_root=Path(f"/tmp/{self._collection_group}"),
+            nexus_file_path=Path(f"/tmp{self._collection_group}.nxs"),
+        )
+
+
+def data_writing_wrapper(
+    plan: MsgGenerator,
+    provider: DataCollectionProvider,
+) -> MsgGenerator:
+    staging = False
+    for message in plan:
+        if message.command == "stage":
+            if not staging:
+                yield from bps.wait_for([provider.update])
+                staging = True
+            if provider.current_data_collection is None:
+                raise Exception("There is no active data collection")
+        elif staging:
+            staging = False
+
+        if message.command == "open_run":
+            if provider.current_data_collection is None:
+                yield from bps.wait_for([provider.update])
+            if provider.current_data_collection is None:
+                raise Exception("There is no active data collection")
+            message.kwargs[
+                DATA_COLLECTION_NUMBER
+            ] = provider.current_data_collection.collection_number
+        yield message
+
+
+data_writing_decorator = make_decorator(data_writing_wrapper)
```

Comment:
This is the only actual part of this file that should stick around. I'm not convinced it should be called `plugins/data_writing.py` either. Maybe `preprocessors/xyz.py`

Reply:
If it's going to be a plan preprocessor should it be moved to dls-bluesky-core?

Reply:
I thought about this. It is specifically a part of blueapi, that fudges plans to write data in our structure, maybe it shouldn't be allowed to get outside of blueapi?

Reply:
I'm fine with it existing only within BlueAPI and being applied on plan import: prevents possibility of attempting to apply multiple times (pretty sure from glancing through the logic that would just doubly increment the data collection number each run, but that's disruptive enough) or being missed. 

Should anyone want to be a. using vanilla Bluesky/not BlueAPI, b. wanting to write to DLS filesystem, they can extract and duplicate the logic. But I don't think that'll be very often.